### ISSUE TITLE
"Go to superproject" button tooltip misleading or missing

### ISSUE THREAD

**Issue by iTob191** [gitextensions/gitextensions#5872]
**Do you want to request a *feature* or report a *bug*?**
Feature

Follow up of #5451 and #5588

**What is the current behavior?**
There are 2 problems with the current behavior of the "Go to superproject" button (![grafik](https://user-images.githubusercontent.com/8921465/49687588-52f50e00-fb05-11e8-9046-a2473e7e892a.png)) in my opinion:

1. If there is no superproject, the button has no tooltip at all. So the only way to find out what this button does, is to click it and see what happens. Although doing so does not trigger any unwanted actions in this case, it's far from ideal.

2. If there is a superproject, the button has the tooltip text "Go to superproject". This is true for the left part of the button, but wrong for the drop-down part on the right.

**Solution**
I'm not really sure what'd be the best solution to this. Two possibilities that came to my mind:
1. One could apply different tooltips to both parts of the button and in case there is no super project disable the left part. But I think this isn't possible with the default implementation of that button and would thereby require a custom implementation which is probably a bit too much overhead for such a small problem. Alternatively, one could describe both parts in the same tooltip, e.g. "Button: Go to superproject, Drop-Down: Go to submodule" but this isn't ideal either in my opinion.
2. Split the functionality into 2 buttons. A normal button on the left for "Go to superproject" that is disabled or hidden when there is no superproject. And a dropdown for "Go to submodule" on the right.

**Environment you encounter the issue:**
- Git Extensions 3.00.00.4433
- Build fca7cf228b481ee8c1b779cf7b882ccdfbdcd1bc
- Git 2.19.2.windows.1
- Microsoft Windows NT 10.0.17134.0
- .NET Framework 4.7.3221.0
- DPI 96dpi (no scaling)

**Comment by gerhardol** [comment]
See also #5669 for proposed changes

**Comment by iTob191** [comment]
That PR sounds interesting but as far as I can tell from the initial post and code, it doesn't touch that button at all so doesn't solve this problem.

> The button icon should be sufficient to show the functionality.
> Can the submodule icons be improved?

For common icons like the gear wheel for settings or the refresh icon that one can find in many different applications, I agree. But submodules are pretty specific to git and I personally can't think of an icon that I would immediately associate with them. That's why I think the current icon is fine, but should always be supplemented by a tooltip.

> The behavior is similar for the other dropdown buttons

My main concern isn't the drop down part of the ToolStripSplitButton. When I see an arrow like the one on the right of the ToolStripSplitButton, I expect a drop-down to open and even if there is no tooltip I can usually safely click that part and just see what kind of drop-down will open.
On the other hand, the left part of a ToolStripSplitButton is a "normal" button, so I have to expect some kind of action to be triggered if I click it. And probably nobody wants to trigger an unknown action on his git repository and just see what happens.

> The tooltip cannot explain everything...

I agree that solving my second point is probably too much for a tooltip.

So instead of showing no tooltip if there is no superproject, I would suggest showing some text like "Submodules...".

**Comment by RussKie** [comment]
> So instead of showing no tooltip if there is no superproject, I would suggest showing some text like "Submodules...".

Just to clarify, the proposal is to show text on the button, right? If so, I don't particularly like it as it sits in the prime real estate.

**Comment by drewnoakes** [comment]
@iTob191 I agree there's room to improve that button and there's been some discussion about it already. Please do take a look at #5717 and #5704 and the discussions, and share your thoughts. I'd like to have this fixed for the 3.1 release.

> When I see an arrow like the one on the right of the ToolStripSplitButton, I expect a drop-down to open and even if there is no tooltip I can usually safely click that part and just see what kind of drop-down will open.

I agree that the split button feels a bit like a booby trap :) 

I strongly believe we should split 'go to superproject' out into a separate button that's only visible when there is a superproject, has a different icon, and always has a tooltip.

IIRC in #5717 the 'submodules' button sometimes goes up, sometimes manages, which is a terrible user experience.

**Comment by iTob191** [comment]
> I interpret the outcome of the discussion to have a tooltip "Submodules..." when there are no superproject (and the button opens the drop-down).

That's correct.

> The PR I intended to link was #5717
That changes the icon back to 2.x if in a submodule

> I agree that the split button feels a bit like a booby trap :)
I strongly believe we should split 'go to superproject' out into a separate button that's only visible when there is a superproject, has a different icon, and always has a tooltip.

I'm not sure which approach I'd prefer. The latter is clearer but takes up space for 2 buttons while the former only for 1. On the other hand, if each of those 2 buttons is only visible if there really is a superproject or submodule, then the latter solution wouldn't take up any space for the majority of repositories anyway. And in case that both, a superproject and submodules, exist, some additional verbosity might be good. Given that, I think I favor the latter.
Anyway, no matter what approach you'll decide on, my issue will be solved by both of them.

> IIRC in #5717 the 'submodules' button sometimes goes up, sometimes manages, which is a terrible user experience.

I agree.

> Anyway, I see this a a refinement of improvements in 3.00, not a regression

That's definitely true. That's why I marked this issue as "Feature".

**Comment by drewnoakes** [comment]
> If the behavior is dynamic, the button may be added at a late stage which is not a good behavior too

I also don't like dancing buttons.

For a _supermodule_, we detect it synchronously in the `GitModule` constructor.

For _submodules_, I'm less sure and would need to do some testing. However it should be possible to identify their existence "immediately" based on the presence or absence of a `.gitmodules` file in the working dir (I think).
